File: tools/objc-api-jsc-diff.py
#!/usr/bin/env python3
"""Compile one Objective-C fixture against zig-js and system JSC, then compare."""

from __future__ import annotations

import difflib
import hashlib
import platform
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
FIXTURE = ROOT / "tests/objc_api_value_diff.m"


def run(command: list[str]) -> bytes:
    return subprocess.run(command, cwd=ROOT, check=True, stdout=subprocess.PIPE).stdout


def compile_fixture(output: Path, extra: list[str]) -> None:
    subprocess.run(
        [
            "xcrun",
            "--sdk",
            "macosx",
            "clang",
            "-fobjc-arc",
            "-fblocks",
            str(FIXTURE),
            *extra,
            "-framework",
            "Foundation",
            "-o",
            str(output),
        ],
        cwd=ROOT,
        check=True,
    )


def main() -> None:
    if len(sys.argv) != 2:
        raise SystemExit("usage: objc-api-jsc-diff.py <libzig-js.a>")
    if platform.system() != "Darwin":
        raise SystemExit("the pinned Objective-C differential gate requires macOS")

    sdk_root = run(["xcrun", "--sdk", "macosx", "--show-sdk-path"]).decode().strip()
    subprocess.run(
        [sys.executable, "tools/verify-objc-api.py", "--sdk-root", sdk_root],
        cwd=ROOT,
        check=True,
    )

    with tempfile.TemporaryDirectory(prefix="zig-js-objc-diff-") as directory:
        temporary = Path(directory)
        reference = temporary / "system-jsc"
        actual = temporary / "zig-js"
        compile_fixture(reference, ["-framework", "JavaScriptCore"])
        compile_fixture(actual, ["-I", str(ROOT / "include"), sys.argv[1], "-lffi"])
        expected_output = run([str(reference)])
        actual_output = run([str(actual)])

    if actual_output != expected_output:
        difference = difflib.unified_diff(
            expected_output.decode().splitlines(),
            actual_output.decode().splitlines(),
            fromfile="system JavaScriptCore",
            tofile="zig-js",
            lineterm="",
        )
        raise SystemExit("Objective-C API differential mismatch:\n" + "\n".join(difference))

    digest = hashlib.sha256(actual_output).hexdigest()[:16]
    print(f"Objective-C JSC differential: matched {len(actual_output.splitlines())} rows ({digest})")


if __name__ == "__main__":
    main()

File: tools/objc-api-jsc-diff.ts
/** Compile the Objective-C value fixture against zig-js and system JSC. */
import { checked, removeTemporaryDirectory, run, sha256File, temporaryDirectory, writeText } from "./lib/home";
declare const __dirname: string;
declare function require(specifier: string): any;
const { verifyObjCAPI } = require("./verify-objc-api");
// The imported verifier is the executable gate at tools/verify-objc-api.ts.
const ROOT = __dirname === "tools" ? "." : __dirname.slice(0, __dirname.lastIndexOf("/tools"));
const join = (left: string, right: string): string => `${left.replace(/\/$/, "")}/${right}`;
const args = process.argv.slice(2);
if (args.length !== 1) throw new Error("usage: objc-api-jsc-diff.ts <libzig-js.a>");
if (checked(["uname", "-s"], "detect platform").trim() !== "Darwin") throw new Error("the pinned Objective-C differential gate requires macOS");
const sdkRoot = checked(["xcrun", "--sdk", "macosx", "--show-sdk-path"], "locate macOS SDK").trim();
verifyObjCAPI(sdkRoot);
const directory = temporaryDirectory("zig-js-objc-diff"), fixture = join(ROOT, "tests/objc_api_value_diff.m");
try {
  const reference = join(directory, "system-jsc"), actual = join(directory, "zig-js");
  checked(["xcrun", "--sdk", "macosx", "clang", "-fobjc-arc", "-fblocks", fixture, "-framework", "JavaScriptCore", "-framework", "Foundation", "-o", reference], "compile Objective-C system JSC fixture");
  checked(["xcrun", "--sdk", "macosx", "clang", "-fobjc-arc", "-fblocks", fixture, "-I", join(ROOT, "include"), args[0], "-framework", "Foundation", "-o", actual], "compile Objective-C zig-js fixture");
  const expectedOutput = checked([reference], `run ${reference}`), actualOutput = checked([actual], `run ${actual}`);
  if (actualOutput !== expectedOutput) {
    const expectedPath = join(directory, "system-jsc.txt"), actualPath = join(directory, "zig-js.txt"); writeText(expectedPath, expectedOutput); writeText(actualPath, actualOutput);
    const difference = run(["diff", "-u", expectedPath, actualPath]); throw new Error(`Objective-C API differential mismatch:\n${difference.stdout || difference.stderr}`);
  }
  const output = join(directory, "actual.txt"); writeText(output, actualOutput);
  const rows = actualOutput ? actualOutput.replace(/\n$/, "").split("\n").length : 0;
  console.log(`Objective-C JSC differential: matched ${rows} rows (${sha256File(output).slice(0, 16)})`);
} finally { removeTemporaryDirectory(directory); }
